### PR TITLE
Fixed code gen for Value Type CS_OUT params

### PR DESCRIPTION
**Pull request by genuinelucifer**
Also polished (sort of) a previous test written by me.
And added an entry in .gitignore to ignore temp files.
.
Also I could not move the function definitions to .cpp file else the CLI part could not find the definition of these functions.

### CODE REVIEW

**Comment by genuinelucifer** [comment]
@tritao @ddobrev 
Pls review.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 1]
```diff
@@ -2616,10 +2616,10 @@ public struct ParamMarshal
                 var paramType = param.Type;
 
                 Class @class;
-                if ( (paramType.GetFinalPointee() ?? paramType).Desugar().TryGetClass(out @class)
-                      && @class.IsRefType)
+                if ( (paramType.GetFinalPointee() ?? paramType).Desugar().TryGetClass(out @class))
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: A redundant space  - '( ('.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 2]
```diff
@@ -2616,10 +2616,10 @@ public struct ParamMarshal
...
                 {
-                    WriteLine("{0} = new {1}();", param.Name, paramType);
+                    if(@class.IsRefType || @class.IsValueType)
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: A missing space after the 'if'. Also, doesn't this check actually cover all cases and is therefore redundant? If it's not, wouldn't !@class.IsInterface be better?

**Comment by ddobrev** [comment]
Looks OK to me save for some minor remarks. However, I am not familiar with this part of C++# so @tritao you should check it out.

**Comment by tritao** [comment]
I think it looks good, but we should be trying to call these new test functions from the NUnit tests. Just making sure they generate and compile doesn't tell us that they actually work when called at runtime.

**Comment by genuinelucifer** [comment]
@tritao Done. I actually thought that generation would gaurantee functioning. But, after thinking for a while it now makes much sense to test them. Thanks.

File: src/Generator/Generators/CSharp/CSharpTextTemplate.cs [hunk 3]
```diff
@@ -2616,8 +2616,8 @@ public struct ParamMarshal
                 var paramType = param.Type;
 
                 Class @class;
-                if ( (paramType.GetFinalPointee() ?? paramType).Desugar().TryGetClass(out @class)
-                      && @class.IsRefType)
+                if ((paramType.GetFinalPointee() ?? paramType).Desugar().TryGetClass(out @class)
+                    && !@class.IsInterface)
```
Comment (tritao) [inline src/Generator/Generators/CSharp/CSharpTextTemplate.cs]: Do we have any test for the `!IsInterface` case?

File: tests/Basic/Basic.Tests.cs [hunk 4]
```diff
@@ -522,5 +522,17 @@ public void TestMarshallingEmptyType()
     {
         var empty = new ReturnsEmpty().Empty;
     }
+
+    [Test]
+    public void TestClassesPassTry()
+    {
+        RefTypeClassPassTry refTypeClassPassTry;
```
Comment (tritao) [inline tests/Basic/Basic.Tests.cs]: You're still not asserting anything. We should have some fields and make sure they're being marshaled correctly.

**Comment by tritao** [comment]
Nevermind about the asserts, I think we're already well-covered. I'd just like to see one test for the interface case.

**Comment by genuinelucifer** [comment]
@tritao I used following in CSharpTemp.h
```C++
void DLL_API funcTryInterfaceTypePtrOut(CS_OUT TestParamToInterfacePassBaseTwo* classTry) { }
void DLL_API funcTryInterfaceTypeOut(CS_OUT TestParamToInterfacePassBaseTwo classTry) { }
```
In the generated code parameter types were changed to respective Interfaces but they were NOT out parameters!
So, if we are not generating out parameters for Interface types then we don't need any checks. So i removed the check ```!@class.IsInterface```
Although, I guess we should generate out parameters for interface types too...
Or is there any limitation/hindrance to this?

**Comment by genuinelucifer** [comment]
@ddobrev @tritao Should i try to generate 'out' qualifier for interfaces too?

**Comment by ddobrev** [comment]
@genuinelucifer yes. Interfaces do not exist on their own, they are added in to support multiple inheritance. This means that existing classes implement those interfaces and if there's any mismatch in signatures, the generated code won't compile.

**Comment by genuinelucifer** [comment]
@tritao @ddobrev I added the proper code for interface types too. Please review the changes. Will submit a new PR (Removing the conflicts) within a couple of hours.

**Comment by genuinelucifer** [comment]
After pulling the recent commits, I have not been able to build the master branch. I am upgrading llvm and clang. And rebuilding everything. Will submit a PR after solving these issues.

**Comment by ddobrev** [comment]
Sorry about that @genuinelucifer, we forgot to tell you. We were given the opportunity to have an OS X build and decided that since we had to build binaries for OS X anyway, we might as well update to the then latest LLVM/Clang.

**Comment by genuinelucifer** [comment]
@ddobrev No problem. The errors revealed themselves that there was some
problem with my llvm. And good to hear about the new OS X build!  :)

**Pull request closed without merging**